timemachine/cpp/src/kernels/k_log_flat_bottom_bond.cuh

```diff
@@ -0,0 +1,112 @@
+#include "../fixed_point.hpp"
+#include "k_fixed_point.cuh"
+#include "k_flat_bottom_bond.cuh"
+
+// branchless implementation of piecewise function
+template <typename RealType>
+RealType __device__ __forceinline__
+compute_log_flat_bottom_energy(RealType k, RealType r, RealType rmin, RealType rmax, RealType beta) {
+    RealType r_gt_rmax = static_cast<RealType>(r > rmax);
+    RealType r_lt_rmin = static_cast<RealType>(r < rmin);
+    RealType nrg = compute_flat_bottom_energy(k, r, rmin, rmax);
+    RealType log_nrg = -log(1 - exp(-beta * nrg)) / beta;
```

Comment:
Haven't thought about this carefully regarding overflow:
(1) do any precautions need to be taken, e.g. https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf
(2) roughly what range of `nrg` will cause overflow at our default fixed-point settings?

-------

Also, nit: naming -- `log_nrg` could be misleading, does this not have units of energy (kJ/mol)?

Reply:
I've implemented equation 7 - I think the paper raises good points, since we do have cases where x=0 (when we're at the flat bottom), and where x is large-ish (where we are hitting the repulsive barrier).

Reply:
To address #2) I don't think we have to worry about overflows for large energies, but forces may be problematic for very small values of energy:

``` cpp

#include <iostream>
#include <cmath>

using namespace std;

#define FIXED_EXPONENT 0x1000000000


template <typename RealType> unsigned long long FLOAT_TO_FIXED_BONDED(RealType v) {
    return static_cast<unsigned long long>(llrintf(v * FIXED_EXPONENT));
}

template <typename RealType> RealType FIXED_TO_FLOAT(unsigned long long v) {
    return static_cast<RealType>(static_cast<long long>(v)) / FIXED_EXPONENT;
}

template <typename RealType> RealType identity(RealType v) {
    return FIXED_TO_FLOAT<RealType>(FLOAT_TO_FIXED_BONDED<RealType>(v));
}



template <typename RealType> RealType stable_log_1_exp_neg(RealType x) {
    const RealType LOG_2 = 0.693147180559945309417232121;
    return x < LOG_2 ? log(-expm1(-x)) : log1p(-exp(-x));
}


template <typename RealType> RealType fn(RealType nrg) {
    RealType beta = 1.0;
    // return -log(1 - exp(-beta * nrg)) / beta;
    return identity(-stable_log_1_exp_neg(beta*nrg)/beta);
}


template <typename RealType> RealType fn_grad_exact(RealType nrg) {
    RealType beta = 1.0;
    return -exp(-beta * nrg) / (1 - exp(-beta * nrg));
}


template <typename RealType> RealType fn_grad_fixed(RealType nrg) {
    return identity(fn_grad_exact(nrg));
}


int main() {
    std::cout << fn(0.0) << " " << fn_grad_exact(0.0) << " " << fn_grad_fixed(0.0) << std::endl;
    std::cout << fn(1e-10) << " " << fn_grad_exact(1e-10) << " " << fn_grad_fixed(1e-10) << std::endl;
    std::cout << fn(1e-9) << " " << fn_grad_exact(1e-9) << " " << fn_grad_fixed(1e-9) << std::endl;
    std::cout << fn(1e-8) << " " << fn_grad_exact(1e-8) << " " << fn_grad_fixed(1e-8) << std::endl;    
    std::cout << fn(1e-7) << " " << fn_grad_exact(1e-7) << " " << fn_grad_fixed(1e-7) << std::endl;    
    std::cout << fn(1e-6) << " " << fn_grad_exact(1e-6) << " " << fn_grad_fixed(1e-6) << std::endl;    
    std::cout << fn(1e-5) << " " << fn_grad_exact(1e-5) << " " << fn_grad_fixed(1e-5) << std::endl;
    std::cout << fn(1e-4) << " " << fn_grad_exact(1e-4) << " " << fn_grad_fixed(1e-4) << std::endl;
    std::cout << fn(1e-3) << " " << fn_grad_exact(1e-3) << " " << fn_grad_fixed(1e-3) << std::endl;
    std::cout << fn(1e-2) << " " << fn_grad_exact(1e-2) << " " << fn_grad_fixed(1e-2) << std::endl;
    std::cout << fn(1e-1) << " " << fn_grad_exact(1e-1) << " " << fn_grad_fixed(1e-1) << std::endl;
    std::cout << fn(1.0) << " " << fn_grad_exact(1.0) << " " << fn_grad_fixed(1.0) << std::endl;
    std::cout << fn(10.0) << " " << fn_grad_exact(10.0) << " " << fn_grad_fixed(10.0) << std::endl;
    std::cout << fn(100.0) << " " << fn_grad_exact(100.0) << " " << fn_grad_fixed(100.0) << std::endl;
    std::cout << fn(1000.0) << " " << fn_grad_exact(1000.0) << " " << fn_grad_fixed(1000.0) << std::endl;
}
```

Results in:

```
-1.34218e+08 -inf -1.34218e+08
23.0259 -1e+10 -1.34218e+08
20.7233 -1e+09 -1.34218e+08 <-- overflow
18.4207 -1e+08 -1e+08
16.1181 -1e+07 -1e+07
13.8155 -1e+06 -1e+06
11.5129 -99999.5 -99999.5
9.21039 -9999.5 -9999.5
6.90826 -999.5 -999.5
4.61017 -99.5008 -99.5008
2.35217 -9.50833 -9.50833
0.458675 -0.581977 -0.581977
4.5401e-05 -4.5402e-05 -4.5402e-05
0 -3.72008e-44 0
0 -0 0
```

Reply:
Thanks!

>I don't think we have to worry about overflows for large energies, but forces may be problematic for very small values of energy:

That looks perfect to me, even at small input values.

(Spot-on down to 1e-8 inputs. At even smaller inputs, gets the right sign, big magnitude. I can't think of an application that would be sensitive to the difference between -inf and -1e+8.)

Reply:
> gets the right sign

I can't guarantee this for *all* inputs, just the ones I've checked. But I'm not as worried here, because we're not evaluates samples generated without the restraints using hamiltonians *including* the restraint (unlike the nonbonded kernels). If this becomes a problem later on I'll be happy to do more sensitivity analysis, I think this is sufficient for now.